modules/007-registrypackages/images/d8-shutdown-inhibitor/src/pkg/inputdev/watcher.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2025 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package inputdev
+
+import (
+	"fmt"
+	"syscall"
+	"unsafe"
+)
+
+type Watcher struct {
+	devs      []Device
+	buttons   []Button
+	stopCh    chan struct{}
+	pressedCh chan struct{}
+}
+
+func NewWatcher(devs []Device, buttons ...Button) *Watcher {
+	return &Watcher{
+		devs:      devs,
+		buttons:   buttons,
+		stopCh:    make(chan struct{}),
+		pressedCh: make(chan struct{}),
+	}
+}
+
+func (w *Watcher) Start() {
+	go w.watch()
+}
+
+func (w *Watcher) Pressed() <-chan struct{} {
+	return w.pressedCh
+}
+
+func (w *Watcher) Stop() {
+	close(w.stopCh)
+}
+
+func (w *Watcher) watch() {
+	evCh := make(chan *inputEvent)
+	go w.readEvents(evCh)
+	for {
+		select {
+		case event := <-evCh:
+			if event.isAnyButtonPressed(w.buttons) {
+				w.pressedCh <- struct{}{}
+			}
+		case <-w.stopCh:
+			return
+		}
+	}
+}
+
+func (w *Watcher) readEvents(evCh chan *inputEvent) {
+	fds := []int{}
+
+	// Open each device.
+	for _, dev := range w.devs {
+		fd, err := syscall.Open(dev.DevPath, syscall.O_RDONLY, 0)
+		if err != nil {
+			continue
+		}
+		defer syscall.Close(fd)
+		fds = append(fds, fd)
+		//devices[fd] = devicePath
+	}
+
+	// Create FdSet bitmask
+	fdSet := syscall.FdSet{}
+	// Max fd to check for Select.
+	fdMax := fds[len(fds)-1] + 1
+
+	// Read events until stopped via channel.
+	for {
+		// Return if watcher was stopped.
+		select {
+		case <-w.stopCh:
+			return
+		default:
```

Comment:
do we need sleep here? for 100 ms for example

Reply:
because we use none block select here

Reply:
It blocks later by the call to `syscall.Select`